tools/clang/unittests/HLSL/WEXAdapter.h

```diff
@@ -0,0 +1,137 @@
+#ifndef LLVM_CLANG_UNITTESTS_WEX_ADAPTER_H
+#define LLVM_CLANG_UNITTESTS_WEX_ADAPTER_H
+
+#ifndef _WIN32
+
+#include <wchar.h>
+#include <unistd.h>
+
+#include "dxc/Support/WinAdapter.h"
+#include "dxc/Support/WinFunctions.h"
+#include "gtest/gtest.h"
+
+#define MAX_PATH 260
+
+// Concatinate two macro fragments
+#define CONCAT2(a, b) a##b
+#define CONCAT1(a, b) CONCAT2(a,b)
+#define CONCAT(a, b) CONCAT1(a,b)
+
+// Determine how many arguments are passed to NARG() up to 3
+#define ARG_CT(_1, _2, _3, N, ...) N
+#define NARG(...) ARG_CT(__VA_ARGS__, 3, 2, 1, 0)
+
+// Call the appropriate arity macro based on number of arguments
+#define CALL_MACRO_N_(PREFIX, N, ...) CONCAT(PREFIX, N)(__VA_ARGS__)
+#define CALL_MACRO_N(PREFIX, ...) CALL_MACRO_N_(PREFIX, NARG(__VA_ARGS__), __VA_ARGS__)
+
+// single and double argument versions
+#define VERIFY_SUCCEEDED_1(expr) EXPECT_TRUE(SUCCEEDED(expr))
+#define VERIFY_SUCCEEDED_2(expr, msg) EXPECT_TRUE(SUCCEEDED(expr)) << msg
+#define VERIFY_SUCCEEDED(...) CALL_MACRO_N(VERIFY_SUCCEEDED_, __VA_ARGS__)
+
+#define VERIFY_ARE_EQUAL_2(A, B) EXPECT_EQ(A, B)
+#define VERIFY_ARE_EQUAL_3(A, B, msg) EXPECT_EQ(A, B) << msg
+#define VERIFY_ARE_EQUAL(...) CALL_MACRO_N(VERIFY_ARE_EQUAL_, __VA_ARGS__)
+
+
+// Macros to convert TAEF macros to gtest equivalents
+#define VERIFY_IS_TRUE EXPECT_TRUE
+#define VERIFY_FAILED(expr) EXPECT_FALSE(SUCCEEDED(expr))
+#define VERIFY_IS_FALSE EXPECT_FALSE
+#define VERIFY_IS_NULL(exp) EXPECT_EQ(nullptr, (exp))
+#define VERIFY_IS_NOT_NULL(exp) EXPECT_NE(nullptr, (exp))
+#define VERIFY_FAIL ADD_FAILURE
+#define VERIFY_ARE_NOT_EQUAL EXPECT_NE
+#define VERIFY_WIN32_BOOL_SUCCEEDED EXPECT_TRUE
+
+#define TEST_CLASS_SETUP(method) bool method(); virtual void SetUp() { EXPECT_TRUE(method()); }
+#define TEST_CLASS_CLEANUP(method) bool method(); virtual void TearDown() { EXPECT_TRUE(method()); }
+#define BEGIN_TEST_CLASS(test)
+#define TEST_CLASS_PROPERTY(str1, str2)
+#define TEST_METHOD_PROPERTY(str1, str2)
+#define END_TEST_CLASS()
+#define TEST_METHOD(method)
+#define BEGIN_TEST_METHOD(method)
+#define END_TEST_METHOD()
+
+// gtest lacks any module setup/cleanup. These functions are called by the main() function
+// before and after tests are run. This approximates the behavior.
+bool moduleSetup();
+bool moduleTeardown();
+#define MODULE_SETUP(method) bool method(); bool moduleSetup() { return method(); }
+#define MODULE_CLEANUP(method) bool method(); bool moduleTeardown() { return method(); }
+
+// No need to expand env vars on Unix platforms, so convert the slashes instead.
+inline DWORD ExpandEnvironmentStringsW(_In_ LPCWSTR lpSrc,_Out_opt_ LPWSTR lpDst, _In_ DWORD nSize) {
```

Comment:
Is this function trying to convert all `\` to `/` ?

Reply:
Would you add to the comment above an example of what an input and output of the function looks like?

Reply:
The reason I'm confused about this method is that the loop below does not increment `lpDst` in every iteration (e.g. if `*lpSrc` is `\` but `wasSlash` is `true`). Was this intended?

Reply:
This function converts backslashes, more specifically, double backslashes, to slashes. There are a few cases where we have sequences of slashes or backslashes. This coalesces them by only incrementing lpDst, when there is something new and not just another slash or backslash. Since I was converting backslashes to slashes and coalescing two anyway, coalescing them all was simpler and always resulted in the same behavior in a Unix style pathname.